Assert: unordered/block/unordered with leaf at same level as unordered causes assert.

- test_nested_three_unordered_block_ordered_with_blank_atx_a
- test_nested_three_unordered_block_ordered_with_blank_text_atx_x
- test_nested_three_unordered_block_ordered_with_blank_thematic_a
- test_nested_three_unordered_block_ordered_with_blank_fenced_a
- test_nested_three_unordered_block_ordered_with_blank_html_a

```
            POGGER.debug("current_indent_level>>:$:<", current_indent_level)
            current_indent_level -= xposition_marker.index_indent
            POGGER.debug("current_indent_level>>:$:<", current_indent_level)
>           assert current_indent_level >= 0
E           AssertionError

pymarkdown\container_blocks\container_block_leaf_processor.py:881: AssertionError
```

--- a/changelog.md
+++ b/changelog.md
@@ -93,6 +93,10 @@ involved tab characters and containers.
 - [Issue 878](https://github.com/jackdewinter/pymarkdown/issues/878)
   - a doubly indented list with a new paragraph continuation line starting
     with multiple tabs was not capturing the first tab properly in the list
+- [Issue 888](https://github.com/jackdewinter/pymarkdown/issues/888)
+  - during fixing of 731, found some outside cases which were throwing asserts
+    as they were outside of normal paths
+  - mostly dealt with proper spacing with block quotes nested within lists.
 
 ## Version 0.9.14 - Date: 2023-10-31
 

--- a/publish/coverage.json
+++ b/publish/coverage.json
@@ -2,12 +2,12 @@
     "projectName": "pymarkdown",
     "reportSource": "pytest",
     "branchLevel": {
-        "totalMeasured": 4267,
-        "totalCovered": 4267
+        "totalMeasured": 4297,
+        "totalCovered": 4297
     },
     "lineLevel": {
-        "totalMeasured": 17873,
-        "totalCovered": 17873
+        "totalMeasured": 17932,
+        "totalCovered": 17932
     }
 }
 

--- a/publish/pylint_suppression.json
+++ b/publish/pylint_suppression.json
@@ -31,7 +31,7 @@
         },
         "pymarkdown/container_blocks/container_block_leaf_processor.py": {
             "too-few-public-methods": 1,
-            "too-many-arguments": 5,
+            "too-many-arguments": 6,
             "chained-comparison": 1,
             "too-many-boolean-expressions": 1
         },
@@ -51,7 +51,8 @@
             "too-many-public-methods": 1
         },
         "pymarkdown/container_blocks/container_helper.py": {
-            "too-few-public-methods": 1
+            "too-few-public-methods": 1,
+            "too-many-arguments": 2
         },
         "pymarkdown/container_blocks/container_indices.py": {},
         "pymarkdown/container_blocks/parse_block_pass_properties.py": {},
@@ -117,7 +118,8 @@
             "too-many-arguments": 5
         },
         "pymarkdown/html/html_helper.py": {
-            "too-many-arguments": 3
+            "too-many-arguments": 4,
+            "too-many-locals": 1
         },
         "pymarkdown/html/html_raw_helper.py": {},
         "pymarkdown/inline/emphasis_helper.py": {
@@ -157,10 +159,11 @@
             "too-many-locals": 5
         },
         "pymarkdown/leaf_blocks/atx_leaf_block_processor.py": {
-            "too-many-arguments": 4
+            "too-many-arguments": 5,
+            "too-many-locals": 2
         },
         "pymarkdown/leaf_blocks/fenced_leaf_block_processor.py": {
-            "too-many-arguments": 11,
+            "too-many-arguments": 12,
             "too-many-locals": 6
         },
         "pymarkdown/leaf_blocks/indented_leaf_block_processor.py": {
@@ -477,8 +480,8 @@
         "too-many-instance-attributes": 23,
         "too-many-public-methods": 4,
         "too-few-public-methods": 40,
-        "too-many-arguments": 205,
-        "too-many-locals": 27,
+        "too-many-arguments": 211,
+        "too-many-locals": 30,
         "chained-comparison": 1,
         "too-many-boolean-expressions": 2,
         "protected-access": 23,

--- a/publish/test-results.json
+++ b/publish/test-results.json
@@ -932,10 +932,10 @@
         },
         {
             "name": "test.nested_three.test_markdown_nested_three_unordered_block_ordered",
-            "totalTests": 42,
+            "totalTests": 44,
             "failedTests": 0,
             "errorTests": 0,
-            "skippedTests": 5,
+            "skippedTests": 3,
             "elapsedTimeInMilliseconds": 0
         },
         {

--- a/pymarkdown/container_blocks/container_block_leaf_processor.py
+++ b/pymarkdown/container_blocks/container_block_leaf_processor.py
@@ -228,6 +228,7 @@ class ContainerBlockLeafProcessor:
                     leaf_token_whitespace,
                     grab_bag.block_quote_data,
                     grab_bag.original_line,
+                    grab_bag,
                 )
                 or IndentedLeafBlockProcessor.parse_indented_code_block(
                     parser_state,
@@ -250,6 +251,7 @@ class ContainerBlockLeafProcessor:
                     leaf_token_whitespace,
                     grab_bag.block_quote_data,
                     grab_bag.original_line,
+                    grab_bag,
                 )
                 or LeafBlockProcessorParagraph.parse_paragraph(
                     parser_state,
@@ -305,6 +307,7 @@ class ContainerBlockLeafProcessor:
             grab_bag.original_line,
             detabified_original_start_index,
             grab_bag.block_quote_data,
+            grab_bag,
         )
 
         ignore_lrd_start = (
@@ -900,6 +903,50 @@ class ContainerBlockLeafProcessor:
 
     # pylint: disable=too-many-arguments
     @staticmethod
+    def __calculate_current_indent_level_loop_kludge(
+        parser_state: ParserState,
+        total_ws: int,
+        current_indent_level: int,
+        last_list_index: int,
+        current_stack_index: int,
+        text_removed_by_container: Optional[str],
+        non_last_block_index: int,
+        last_block_index: int,
+        line_number: int,
+    ) -> Tuple[bool, int, Optional[int], int]:
+        # TODO THIS IS A KLUDGE
+        #
+        # As soon as we go past here, if we are going to get rid of the
+        keep_processing = True
+        new_indent_level = None
+        if (
+            last_list_index > 0
+            and text_removed_by_container is None
+            and len(parser_state.token_stack) - 1 == current_stack_index
+        ):
+            keep_processing = total_ws != current_indent_level
+
+        if keep_processing:
+            last_list_index = 0
+            (
+                new_indent_level,
+                non_last_block_index,
+            ) = ContainerBlockLeafProcessor.__calculate_current_indent_level_block_quote(
+                parser_state,
+                current_stack_index,
+                non_last_block_index,
+                last_block_index,
+                line_number,
+                current_indent_level,
+                text_removed_by_container,
+            )
+            keep_processing = new_indent_level is not None
+        return keep_processing, last_list_index, new_indent_level, non_last_block_index
+
+    # pylint: enable=too-many-arguments
+
+    # pylint: disable=too-many-arguments
+    @staticmethod
     def __calculate_current_indent_level_loop(
         parser_state: ParserState,
         last_block_index: int,
@@ -918,20 +965,22 @@ class ContainerBlockLeafProcessor:
         continue_in_loop = True
         keep_processing = True
         if parser_state.token_stack[current_stack_index].is_block_quote:
-            last_list_index = 0
             (
+                keep_processing,
+                last_list_index,
                 new_indent_level,
                 non_last_block_index,
-            ) = ContainerBlockLeafProcessor.__calculate_current_indent_level_block_quote(
+            ) = ContainerBlockLeafProcessor.__calculate_current_indent_level_loop_kludge(
                 parser_state,
+                total_ws,
+                current_indent_level,
+                last_list_index,
                 current_stack_index,
+                text_removed_by_container,
                 non_last_block_index,
                 last_block_index,
                 line_number,
-                current_indent_level,
-                text_removed_by_container,
             )
-            keep_processing = new_indent_level is not None
             if keep_processing:
                 proposed_indent_level = new_indent_level
         elif parser_state.token_stack[current_stack_index].is_list:

--- a/pymarkdown/container_blocks/container_helper.py
+++ b/pymarkdown/container_blocks/container_helper.py
@@ -9,8 +9,10 @@ import logging
 from typing import List, Optional, cast
 
 from pymarkdown.block_quotes.block_quote_data import BlockQuoteData
+from pymarkdown.container_blocks.container_grab_bag import ContainerGrabBag
 from pymarkdown.general.parser_logger import ParserLogger
 from pymarkdown.general.parser_state import ParserState
+from pymarkdown.general.position_marker import PositionMarker
 from pymarkdown.tokens.block_quote_markdown_token import BlockQuoteMarkdownToken
 from pymarkdown.tokens.markdown_token import EndMarkdownToken, MarkdownToken
 from pymarkdown.tokens.stack_token import ListStackToken
@@ -28,28 +30,28 @@ class ContainerHelper:
     @staticmethod
     def __reduce_containers_if_required_bq_list(
         parser_state: ParserState,
+        position_marker: PositionMarker,
         extracted_whitespace: Optional[str],
         new_tokens: List[MarkdownToken],
     ) -> bool:
         did_once = False
         if extracted_whitespace is not None and parser_state.token_stack[-1].is_list:
             search_index = len(parser_state.token_stack)
-            leading_space_length = len(extracted_whitespace)
+            leading_space_length = (
+                len(extracted_whitespace) + position_marker.index_indent
+            )
             while parser_state.token_stack[search_index - 1].is_list:
                 list_token = cast(
                     ListStackToken, parser_state.token_stack[search_index - 1]
                 )
-                assert list_token.indent_level > leading_space_length
-                # if list_token.indent_level <= leading_space_length:
-                #     break
+                if list_token.indent_level <= leading_space_length:
+                    break
                 search_index -= 1
                 did_once = True
 
-            POGGER.debug("lsl $", parser_state.token_stack[search_index])
-
-            # if did_once:
-            last_token = cast(EndMarkdownToken, new_tokens[-1])
-            last_token.set_extra_end_data(None)
+            if did_once:
+                last_token = cast(EndMarkdownToken, new_tokens[-1])
+                last_token.set_extra_end_data(None)
 
             (
                 container_level_tokens,
@@ -64,12 +66,15 @@ class ContainerHelper:
             new_tokens.extend(container_level_tokens)
         return did_once
 
+    # pylint: disable=too-many-arguments
     @staticmethod
     def __reduce_containers_if_required_bq(
         parser_state: ParserState,
+        position_marker: PositionMarker,
         new_tokens: List[MarkdownToken],
         split_tab: bool,
         extracted_whitespace: Optional[str],
+        grab_bag: ContainerGrabBag,
     ) -> bool:
         # TODO cyclic?
         x_tokens, _ = parser_state.close_open_blocks_fn(
@@ -83,8 +88,15 @@ class ContainerHelper:
         first_new_token = cast(EndMarkdownToken, x_tokens[0])
 
         did_reduce_list = ContainerHelper.__reduce_containers_if_required_bq_list(
-            parser_state, extracted_whitespace, x_tokens
+            parser_state, position_marker, extracted_whitespace, x_tokens
         )
+        was_list_ended = False
+        if (
+            len(grab_bag.container_tokens)
+            and grab_bag.container_tokens[-1].is_end_token
+        ):
+            end_token = cast(EndMarkdownToken, grab_bag.container_tokens[-1])
+            was_list_ended = end_token.start_markdown_token.is_list_start
 
         matching_start_token = cast(
             BlockQuoteMarkdownToken, first_new_token.start_markdown_token
@@ -112,7 +124,7 @@ class ContainerHelper:
         POGGER.debug("extra_end_data>>:$:<", first_new_token.extra_end_data)
         assert first_new_token.extra_end_data is None
 
-        if did_reduce_list:
+        if did_reduce_list or was_list_ended:
             first_new_token.set_extra_end_data(None)
         else:
             first_new_token.set_extra_end_data(last_newline_part)
@@ -120,13 +132,18 @@ class ContainerHelper:
         new_tokens.extend(x_tokens)
         return split_tab
 
+    # pylint: enable=too-many-arguments
+
+    # pylint: disable=too-many-arguments
     @staticmethod
     def reduce_containers_if_required(
         parser_state: ParserState,
+        position_marker: PositionMarker,
         block_quote_data: BlockQuoteData,
         new_tokens: List[MarkdownToken],
         split_tab: bool,
-        extracted_whitespace: Optional[str] = None,
+        extracted_whitespace: Optional[str],
+        grab_bag: ContainerGrabBag,
     ) -> bool:
         """
         Given a drop in the current count of block quotes versus what is actually
@@ -149,7 +166,12 @@ class ContainerHelper:
             and parser_state.token_stack[-1].is_block_quote
         ):
             split_tab = ContainerHelper.__reduce_containers_if_required_bq(
-                parser_state, new_tokens, split_tab, extracted_whitespace
+                parser_state,
+                position_marker,
+                new_tokens,
+                split_tab,
+                extracted_whitespace,
+                grab_bag,
             )
 
         # if did_close_bq and extracted_whitespace is not None and parser_state.token_stack[-1].is_list:
@@ -183,6 +205,8 @@ class ContainerHelper:
         #         )
         #         new_tokens.extend(container_level_tokens)
         return split_tab
+
+    # pylint: enable=too-many-arguments
 
 
 # pylint: enable=too-few-public-methods

--- a/pymarkdown/html/html_helper.py
+++ b/pymarkdown/html/html_helper.py
@@ -6,6 +6,7 @@ import string
 from typing import List, Optional, Tuple, cast
 
 from pymarkdown.block_quotes.block_quote_data import BlockQuoteData
+from pymarkdown.container_blocks.container_grab_bag import ContainerGrabBag
 from pymarkdown.container_blocks.container_helper import ContainerHelper
 from pymarkdown.container_blocks.parse_block_pass_properties import (
     ParseBlockPassProperties,
@@ -594,7 +595,7 @@ class HtmlHelper:
             removed_chars_at_start = len(original_prefix)
         return alternate_list_leading_space, removed_chars_at_start
 
-    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-arguments, too-many-locals
     @staticmethod
     def __found_html_block(
         parser_state: ParserState,
@@ -604,6 +605,7 @@ class HtmlHelper:
         extracted_whitespace: Optional[str],
         block_quote_data: BlockQuoteData,
         html_block_type: str,
+        grab_bag: ContainerGrabBag,
     ) -> Tuple[List[MarkdownToken], bool, Optional[int]]:
         split_tab = False
         alternate_list_leading_space: Optional[str] = None
@@ -638,7 +640,13 @@ class HtmlHelper:
         old_split_tab = split_tab
         did_adjust_block_quote = False
         if split_tab := ContainerHelper.reduce_containers_if_required(
-            parser_state, block_quote_data, new_tokens, split_tab, extracted_whitespace
+            parser_state,
+            position_marker,
+            block_quote_data,
+            new_tokens,
+            split_tab,
+            extracted_whitespace,
+            grab_bag,
         ):
             TabHelper.adjust_block_quote_indent_for_tab(
                 parser_state,
@@ -657,8 +665,9 @@ class HtmlHelper:
         parser_state.token_stack.append(HtmlBlockStackToken(html_block_type, new_token))
         return new_tokens, did_adjust_block_quote, removed_chars_at_start
 
-    # pylint: enable=too-many-arguments
+    # pylint: enable=too-many-arguments, too-many-locals
 
+    # pylint: disable=too-many-arguments
     @staticmethod
     def parse_html_block(
         parser_state: ParserState,
@@ -666,6 +675,7 @@ class HtmlHelper:
         extracted_whitespace: Optional[str],
         block_quote_data: BlockQuoteData,
         original_line: str,
+        grab_bag: ContainerGrabBag,
     ) -> Tuple[List[MarkdownToken], bool, Optional[int]]:
         """
         Determine if we have the criteria that we need to start an HTML block.
@@ -702,11 +712,14 @@ class HtmlHelper:
                 extracted_whitespace,
                 block_quote_data,
                 html_block_type,
+                grab_bag,
             )
         else:
             new_tokens = []
         POGGER.debug("did_adjust_block_quote=$", did_adjust_block_quote)
         return new_tokens, did_adjust_block_quote, removed_chars_at_start
+
+    # pylint: enable=too-many-arguments
 
     @staticmethod
     def check_blank_html_block_end(parser_state: ParserState) -> List[MarkdownToken]:

--- a/pymarkdown/leaf_blocks/atx_leaf_block_processor.py
+++ b/pymarkdown/leaf_blocks/atx_leaf_block_processor.py
@@ -5,6 +5,7 @@ import logging
 from typing import List, Optional, Tuple, cast
 
 from pymarkdown.block_quotes.block_quote_data import BlockQuoteData
+from pymarkdown.container_blocks.container_grab_bag import ContainerGrabBag
 from pymarkdown.container_blocks.container_helper import ContainerHelper
 from pymarkdown.general.parser_helper import ParserHelper
 from pymarkdown.general.parser_logger import ParserLogger
@@ -75,6 +76,7 @@ class AtxLeafBlockProcessor:
                 )
         return False, None, None, None
 
+    # pylint: disable=too-many-arguments
     @staticmethod
     def parse_atx_headings(
         parser_state: ParserState,
@@ -82,6 +84,7 @@ class AtxLeafBlockProcessor:
         extracted_whitespace: Optional[str],
         block_quote_data: BlockQuoteData,
         original_line: str,
+        grab_bag: ContainerGrabBag,
     ) -> List[MarkdownToken]:
         """
         Handle the parsing of an atx heading.
@@ -113,9 +116,12 @@ class AtxLeafBlockProcessor:
             block_quote_data,
             extracted_whitespace_at_start,
             extracted_whitespace,
+            grab_bag,
         )
 
-    # pylint: disable=too-many-arguments
+    # pylint: enable=too-many-arguments
+
+    # pylint: disable=too-many-arguments, too-many-locals
     @staticmethod
     def __parse_atx_heading_found(
         parser_state: ParserState,
@@ -126,6 +132,7 @@ class AtxLeafBlockProcessor:
         block_quote_data: BlockQuoteData,
         extracted_whitespace_at_start: Optional[str],
         extracted_whitespace: Optional[str],
+        grab_bag: ContainerGrabBag,
     ) -> List[MarkdownToken]:
         assert extracted_whitespace_at_start is not None
         POGGER.debug(
@@ -151,6 +158,7 @@ class AtxLeafBlockProcessor:
             extracted_whitespace,
             hash_count,
             block_quote_data,
+            grab_bag,
         )
         assert hash_count is not None
         assert extracted_whitespace is not None
@@ -181,8 +189,10 @@ class AtxLeafBlockProcessor:
         )
         return new_tokens
 
-    # pylint: enable=too-many-arguments
+    # pylint: enable=too-many-arguments, too-many-locals
+
     # pylint: disable=too-many-arguments
+
     @staticmethod
     def __parse_atx_heading_add_tokens(
         parser_state: ParserState,
@@ -349,7 +359,7 @@ class AtxLeafBlockProcessor:
 
     # pylint: enable=too-many-arguments
 
-    # pylint: disable=too-many-arguments
+    # pylint: disable=too-many-arguments, too-many-locals
     @staticmethod
     def __prepare_for_create_atx_heading(
         parser_state: ParserState,
@@ -360,6 +370,7 @@ class AtxLeafBlockProcessor:
         extracted_whitespace: Optional[str],
         hash_count: Optional[int],
         block_quote_data: BlockQuoteData,
+        grab_bag: ContainerGrabBag,
     ) -> Tuple[
         StackToken, str, int, str, str, List[MarkdownToken], str, Optional[str], bool
     ]:
@@ -407,7 +418,13 @@ class AtxLeafBlockProcessor:
         new_tokens, _ = parser_state.close_open_blocks_fn(parser_state)
         POGGER.debug("new_tokens>:$:<", new_tokens)
         if ContainerHelper.reduce_containers_if_required(
-            parser_state, block_quote_data, new_tokens, split_tab, extracted_whitespace
+            parser_state,
+            position_marker,
+            block_quote_data,
+            new_tokens,
+            split_tab,
+            extracted_whitespace,
+            grab_bag,
         ):
             POGGER.debug("extracted_whitespace>:$:<", extracted_whitespace)
             extracted_whitespace = TabHelper.adjust_block_quote_indent_for_tab(
@@ -436,7 +453,8 @@ class AtxLeafBlockProcessor:
             ParserHelper.tab_character in original_line and eligble_for_tab_match_delay,
         )
 
-    # pylint: enable=too-many-arguments
+    # pylint: enable=too-many-arguments, too-many-locals
+
     @staticmethod
     def __determine_eligble_for_tab_match_delay(
         parser_state: ParserState,

--- a/pymarkdown/leaf_blocks/fenced_leaf_block_processor.py
+++ b/pymarkdown/leaf_blocks/fenced_leaf_block_processor.py
@@ -5,6 +5,7 @@ import logging
 from typing import List, Optional, Tuple, cast
 
 from pymarkdown.block_quotes.block_quote_data import BlockQuoteData
+from pymarkdown.container_blocks.container_grab_bag import ContainerGrabBag
 from pymarkdown.container_blocks.container_helper import ContainerHelper
 from pymarkdown.general.constants import Constants
 from pymarkdown.general.parser_helper import ParserHelper
@@ -100,6 +101,7 @@ class FencedLeafBlockProcessor:
         original_line: str,
         detabified_original_start_index: int,
         block_quote_data: BlockQuoteData,
+        grab_bag: ContainerGrabBag,
     ) -> Tuple[List[MarkdownToken], Optional[str]]:
         """
         Handle the parsing of a fenced code block
@@ -152,6 +154,7 @@ class FencedLeafBlockProcessor:
                     new_index,
                     block_quote_data,
                     after_fence_index,
+                    grab_bag,
                 )
         elif parser_state.token_stack[-1].is_fenced_code_block:
             POGGER.debug("parse_fenced_code_block:already in")
@@ -367,6 +370,7 @@ class FencedLeafBlockProcessor:
         new_index: int,
         block_quote_data: BlockQuoteData,
         after_fence_index: int,
+        grab_bag: ContainerGrabBag,
     ) -> List[MarkdownToken]:
         POGGER.debug("pfcb->check")
         new_tokens: List[MarkdownToken] = []
@@ -392,6 +396,7 @@ class FencedLeafBlockProcessor:
                 new_index,
                 block_quote_data,
                 after_fence_index,
+                grab_bag,
             )
 
             POGGER.debug("StackToken-->$<<", parser_state.token_stack[-1])
@@ -539,6 +544,7 @@ class FencedLeafBlockProcessor:
         new_index: int,
         block_quote_data: BlockQuoteData,
         after_fence_index: int,
+        grab_bag: ContainerGrabBag,
     ) -> Tuple[StackToken, List[MarkdownToken], Optional[str]]:
         (
             line_to_parse,
@@ -575,17 +581,21 @@ class FencedLeafBlockProcessor:
             block_quote_data,
             split_tab_whitespace,
             adjusted_corrected_prefix,
+            grab_bag,
         )
 
     # pylint: enable=too-many-arguments, too-many-locals
 
+    # pylint: disable=too-many-arguments
     @staticmethod
     def __add_fenced_tokens_calc(
         parser_state: ParserState,
+        position_marker: PositionMarker,
         split_tab: bool,
         block_quote_data: BlockQuoteData,
         split_tab_whitespace: Optional[str],
         extracted_whitespace: Optional[str],
+        grab_bag: ContainerGrabBag,
     ) -> Tuple[StackToken, List[MarkdownToken], int]:
         old_top_of_stack = parser_state.token_stack[-1]
         new_tokens, _ = parser_state.close_open_blocks_fn(
@@ -594,7 +604,13 @@ class FencedLeafBlockProcessor:
         )
 
         if split_tab := ContainerHelper.reduce_containers_if_required(
-            parser_state, block_quote_data, new_tokens, split_tab, extracted_whitespace
+            parser_state,
+            position_marker,
+            block_quote_data,
+            new_tokens,
+            split_tab,
+            extracted_whitespace,
+            grab_bag,
         ):
             TabHelper.adjust_block_quote_indent_for_tab(
                 parser_state,
@@ -606,6 +622,8 @@ class FencedLeafBlockProcessor:
             whitespace_count_delta = 0
 
         return old_top_of_stack, new_tokens, whitespace_count_delta
+
+    # pylint: enable=too-many-arguments
 
     # pylint: disable=too-many-locals, too-many-arguments
     @staticmethod
@@ -694,6 +712,7 @@ class FencedLeafBlockProcessor:
         block_quote_data: BlockQuoteData,
         split_tab_whitespace: Optional[str],
         adjusted_corrected_prefix: Optional[str],
+        grab_bag: ContainerGrabBag,
     ) -> Tuple[StackToken, List[MarkdownToken], Optional[str]]:
         (
             old_top_of_stack,
@@ -701,10 +720,12 @@ class FencedLeafBlockProcessor:
             whitespace_start_count,
         ) = FencedLeafBlockProcessor.__add_fenced_tokens_calc(
             parser_state,
+            position_marker,
             split_tab,
             block_quote_data,
             split_tab_whitespace,
             extracted_whitespace,
+            grab_bag,
         )
 
         pre_extracted_text, pre_text_after_extracted_text = (
@@ -905,6 +926,7 @@ class FencedLeafBlockProcessor:
         original_line: str,
         detabified_original_start_index: int,
         block_quote_data: BlockQuoteData,
+        grab_bag: ContainerGrabBag,
     ) -> bool:
         """
         Take care of the processing for fenced code blocks.
@@ -924,6 +946,7 @@ class FencedLeafBlockProcessor:
             original_line,
             detabified_original_start_index,
             block_quote_data,
+            grab_bag,
         )
         POGGER.debug(">>leaf_token_whitespace>:$:<", leaf_token_whitespace)
         if fenced_tokens:

--- a/pymarkdown/leaf_blocks/leaf_block_processor.py
+++ b/pymarkdown/leaf_blocks/leaf_block_processor.py
@@ -123,6 +123,7 @@ class LeafBlockProcessor:
                 leaf_token_whitespace,
                 grab_bag.block_quote_data,
                 grab_bag.original_line,
+                grab_bag,
             )
             if html_tokens:
                 POGGER.debug(">>html started>>")

--- a/pymarkdown/leaf_blocks/thematic_leaf_block_processor.py
+++ b/pymarkdown/leaf_blocks/thematic_leaf_block_processor.py
@@ -6,6 +6,7 @@ import logging
 from typing import List, Optional, Tuple
 
 from pymarkdown.block_quotes.block_quote_data import BlockQuoteData
+from pymarkdown.container_blocks.container_grab_bag import ContainerGrabBag
 from pymarkdown.container_blocks.container_helper import ContainerHelper
 from pymarkdown.general.parser_helper import ParserHelper
 from pymarkdown.general.parser_logger import ParserLogger
@@ -90,6 +91,7 @@ class ThematicLeafBlockProcessor:
         extracted_whitespace: Optional[str],
         block_quote_data: BlockQuoteData,
         original_line: str,
+        grab_bag: ContainerGrabBag,
     ) -> List[MarkdownToken]:
         """
         Handle the parsing of a thematic break.
@@ -154,6 +156,7 @@ class ThematicLeafBlockProcessor:
                 block_quote_data,
                 split_tab,
                 split_tab_with_block_quote_suffix,
+                grab_bag,
             )
 
             new_tokens.append(
@@ -183,6 +186,7 @@ class ThematicLeafBlockProcessor:
         block_quote_data: BlockQuoteData,
         split_tab: bool,
         split_tab_with_block_quote_suffix: bool,
+        grab_bag: ContainerGrabBag,
     ) -> None:
         if split_tab and not split_tab_with_block_quote_suffix:
             assert extracted_whitespace is not None
@@ -197,7 +201,13 @@ class ThematicLeafBlockProcessor:
                 token_text,
             )
         elif split_tab := ContainerHelper.reduce_containers_if_required(
-            parser_state, block_quote_data, new_tokens, split_tab, extracted_whitespace
+            parser_state,
+            position_marker,
+            block_quote_data,
+            new_tokens,
+            split_tab,
+            extracted_whitespace,
+            grab_bag,
         ):
             TabHelper.adjust_block_quote_indent_for_tab(parser_state)
 

--- a/pymarkdown/transform_gfm/transform_to_gfm_list_looseness.py
+++ b/pymarkdown/transform_gfm/transform_to_gfm_list_looseness.py
@@ -2,7 +2,7 @@
 Module to handle the calculationof list looseness for the GRM transformer.
 """
 import logging
-from typing import List, Tuple, cast
+from typing import List, Optional, Tuple, cast
 
 from pymarkdown.general.parser_logger import ParserLogger
 from pymarkdown.tokens.list_start_markdown_token import ListStartMarkdownToken
@@ -94,8 +94,12 @@ class TransformToGfmListLooseness:
             check_me = True
         elif current_token.is_block_quote_end:
             POGGER.debug("cll>>end block quote>>$", current_token)
-            stack_count = TransformToGfmListLooseness.__handle_block_quote_end(
-                stack_count
+            (
+                stop_me,
+                is_loose,
+                stack_count,
+            ) = TransformToGfmListLooseness.__handle_block_quote_end(
+                stack_count, actual_tokens, current_token_index, stop_me, is_loose
             )
         elif current_token.is_list_end:
             POGGER.debug("cll>>end list>>$", current_token)
@@ -181,9 +185,70 @@ class TransformToGfmListLooseness:
         return stack_count + 1
 
     @staticmethod
-    def __handle_block_quote_end(stack_count: int) -> int:
+    def __handle_block_quote_end_calc(
+        actual_tokens: List[MarkdownToken], search_index: int, current_token_index: int
+    ) -> Tuple[bool, bool]:
+        stop_me = TransformToGfmListLooseness.__is_token_loose(
+            actual_tokens, search_index + 1, True
+        )
+        if stop_me:
+            blank_token = actual_tokens[search_index]
+            search_index = current_token_index
+            next_token = actual_tokens[search_index]
+            while search_index >= 0 and not next_token.is_block_quote_start:
+                search_index -= 1
+                next_token = actual_tokens[search_index]
+            block_quote_spaces_count = len(next_token.bleading_spaces.split("\n"))  # type: ignore
+            end_block_quote_line = next_token.line_number + block_quote_spaces_count
+            stop_me = not blank_token.line_number < end_block_quote_line
+
+        is_loose = stop_me
+        if stop_me:
+            POGGER.debug("!!!LOOSE-look back at end!!!")
+        return is_loose, stop_me
+
+    @staticmethod
+    def __handle_block_quote_end(
+        stack_count: int,
+        actual_tokens: List[MarkdownToken],
+        current_token_index: int,
+        stop_me: bool,
+        is_loose: bool,
+    ) -> Tuple[bool, bool, int]:
         POGGER.debug(">>block--end>>$", stack_count)
-        return stack_count - 1
+        old_stack_count = stack_count
+        stack_count -= 1
+        if old_stack_count and not stack_count:
+            search_index = current_token_index - 1
+            found_end_token: Optional[MarkdownToken] = actual_tokens[search_index]
+            while found_end_token is not None:
+                found_end_token = None
+                next_token = actual_tokens[search_index]
+                if next_token.is_end_token:
+                    next_end_token = cast(EndMarkdownToken, next_token)
+                    if (
+                        next_end_token.start_markdown_token.is_block_quote_start
+                        or next_end_token.start_markdown_token.is_list_start
+                    ):
+                        found_end_token = next_end_token
+                        search_index -= 1
+
+            new_index = current_token_index + 1
+            keep_checking = (
+                new_index < len(actual_tokens) and actual_tokens[new_index].is_end_token
+            )
+            if keep_checking:
+                end_token = cast(EndMarkdownToken, actual_tokens[new_index])
+                keep_checking = end_token.start_markdown_token.is_list_start
+
+            if not keep_checking:
+                (
+                    is_loose,
+                    stop_me,
+                ) = TransformToGfmListLooseness.__handle_block_quote_end_calc(
+                    actual_tokens, search_index, current_token_index
+                )
+        return stop_me, is_loose, stack_count
 
     @staticmethod
     def __handle_blank_line(
@@ -225,7 +290,7 @@ class TransformToGfmListLooseness:
 
     @staticmethod
     def __is_token_loose(
-        actual_tokens: List[MarkdownToken], current_token_index: int
+        actual_tokens: List[MarkdownToken], current_token_index: int, xx: bool = False
     ) -> bool:
         """
         Check to see if this token inspires looseness.
@@ -244,6 +309,8 @@ class TransformToGfmListLooseness:
                 or actual_tokens[check_index - 1].is_list_start
             ):
                 POGGER.debug("!!!Starting Blank!!!")
+            elif actual_tokens[check_index - 1].is_block_quote_start and xx:
+                POGGER.debug("!!!Starting BQ Blank!!!")
             else:
                 POGGER.debug("!!!LOOSE!!!")
                 return True

--- a/test/gfm/test_markdown_lists.py
+++ b/test/gfm/test_markdown_lists.py
@@ -355,7 +355,7 @@ boffo</p>
 </ul>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens)
 
 
 @pytest.mark.gfm

--- a/test/gfm/test_markdown_whitespace_paragraph.py
+++ b/test/gfm/test_markdown_whitespace_paragraph.py
@@ -536,7 +536,7 @@ ghi</li>
 </ol>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens)
 
 
 @pytest.mark.gfm

--- a/test/nested_three/test_markdown_nested_three_unordered_block_ordered.py
+++ b/test/nested_three/test_markdown_nested_three_unordered_block_ordered.py
@@ -1346,11 +1346,10 @@ Some Text
 <p>Some Text</p>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens)
 
 
 @pytest.mark.gfm
-@pytest.mark.skip
 def test_nested_three_unordered_block_ordered_with_blank_atx_a():
     """
     TBD - from https://github.com/jackdewinter/pymarkdown/issues/731
@@ -1367,7 +1366,7 @@ Some Text
 """
 
     expected_tokens = [
-        "[ulist(1,1):-::2::\n\n]",
+        "[ulist(1,1):-::2::  \n\n]",
         "[para(1,3):]",
         "[text(1,3):Test List:]",
         "[end-para:::True]",
@@ -1382,9 +1381,9 @@ Some Text
         "[end-para:::True]",
         "[BLANK(4,1):]",
         "[end-olist:::True]",
-        "[end-block-quote::  > :True]",
-        "[atx(5,1):2:0:]",
-        "[text(5,4):Headline 2: ]",
+        "[end-block-quote:::True]",
+        "[atx(5,3):2:0:]",
+        "[text(5,6):Headline 2: ]",
         "[end-atx::]",
         "[BLANK(6,1):]",
         "[end-ulist:::True]",
@@ -1394,24 +1393,24 @@ Some Text
         "[BLANK(8,1):]",
     ]
     expected_gfm = """<ul>
-<li>Test List
+<li>
+<p>Test List</p>
 <blockquote>
 <ol>
 <li>Test1</li>
 <li>Test2</li>
 </ol>
 </blockquote>
+<h2>Headline 2</h2>
 </li>
 </ul>
-<h2>Headline 2</h2>
 <p>Some Text</p>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens)
 
 
 @pytest.mark.gfm
-@pytest.mark.skip
 def test_nested_three_unordered_block_ordered_with_blank_text_atx_x():
     """
     TBD - from https://github.com/jackdewinter/pymarkdown/issues/731
@@ -1431,7 +1430,7 @@ Some Text
 """
 
     expected_tokens = [
-        "[ulist(1,1):-::2::\n\n]",
+        "[ulist(1,1):-::2::\n  \n\n]",
         "[para(1,3):]",
         "[text(1,3):Test List:]",
         "[end-para:::True]",
@@ -1448,11 +1447,11 @@ Some Text
         "[BLANK(5,1):]",
         "[end-olist:::True]",
         "[end-block-quote:::False]",
-        "[end-ulist:::True]",
-        "[para(6,1):]",
-        "[text(6,1):this is some text:]",
+        "[para(6,3):]",
+        "[text(6,3):this is some text:]",
         "[end-para:::True]",
         "[BLANK(7,1):]",
+        "[end-ulist:::True]",
         "[atx(8,1):2:0:]",
         "[text(8,4):Headline 2: ]",
         "[end-atx::]",
@@ -1478,7 +1477,7 @@ Some Text
 <p>Some Text</p>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens)
 
 
 @pytest.mark.gfm
@@ -1548,7 +1547,7 @@ Some Text
 <p>Some Text</p>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens)
 
 
 @pytest.mark.gfm
@@ -1612,7 +1611,6 @@ def test_nested_three_unordered_block_ordered_with_blank_thematic_x():
 
 
 @pytest.mark.gfm
-@pytest.mark.skip
 def test_nested_three_unordered_block_ordered_with_blank_thematic_a():
     """
     TBD - from https://github.com/jackdewinter/pymarkdown/issues/731
@@ -1633,7 +1631,7 @@ def test_nested_three_unordered_block_ordered_with_blank_thematic_a():
         "[text(1,3):Headline 1: ]",
         "[end-atx::]",
         "[BLANK(2,1):]",
-        "[ulist(3,1):-::2::\n\n]",
+        "[ulist(3,1):-::2::\n  \n\n]",
         "[para(3,3):]",
         "[text(3,3):Test List:]",
         "[end-para:::True]",
@@ -1650,9 +1648,9 @@ def test_nested_three_unordered_block_ordered_with_blank_thematic_a():
         "[BLANK(7,1):]",
         "[end-olist:::True]",
         "[end-block-quote:::True]",
-        "[end-ulist:::True]",
-        "[tbreak(8,1):-::-----]",
+        "[tbreak(8,3):-::-----]",
         "[BLANK(9,1):]",
+        "[end-ulist:::True]",
     ]
     expected_gfm = """<h1>Headline 1</h1>
 <ul>
@@ -1669,7 +1667,7 @@ def test_nested_three_unordered_block_ordered_with_blank_thematic_a():
 </ul>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
 
 
 @pytest.mark.gfm
@@ -1746,7 +1744,7 @@ def test_nested_three_unordered_block_ordered_with_blank_fenced_a():
 """
 
     expected_tokens = [
-        "[ulist(1,1):-::2::\n]",
+        "[ulist(1,1):-::2::  \n\n  \n  \n]",
         "[para(1,3):]",
         "[text(1,3):Test List:]",
         "[end-para:::True]",
@@ -1762,11 +1760,11 @@ def test_nested_three_unordered_block_ordered_with_blank_fenced_a():
         "[BLANK(4,1):]",
         "[end-olist:::True]",
         "[end-block-quote:::True]",
-        "[end-ulist:::True]",
-        "[fcode-block(5,1):`:3:text:::::]",
+        "[fcode-block(5,3):`:3:text:::::]",
         "[text(6,1):block:]",
         "[end-fcode-block:::3:False]",
         "[BLANK(8,1):]",
+        "[end-ulist:::True]",
     ]
     expected_gfm = """<ul>
 <li>
@@ -1783,7 +1781,123 @@ def test_nested_three_unordered_block_ordered_with_blank_fenced_a():
 </ul>"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+
+
+@pytest.mark.gfm
+@pytest.mark.skip
+def test_nested_three_unordered_block_ordered_with_blank_fenced_b():
+    """
+    TBD - from https://github.com/jackdewinter/pymarkdown/issues/731
+    """
+
+    # Arrange
+    source_markdown = """- Test List
+  > 1) Test1
+  > 2) Test2
+  >
+  ```text
+  block
+  ```
+"""
+
+    expected_tokens = [
+        "[ulist(1,1):-::2::  \n\n  \n  \n]",
+        "[para(1,3):]",
+        "[text(1,3):Test List:]",
+        "[end-para:::True]",
+        "[block-quote(2,3):  :  > \n  > \n  >]",
+        "[olist(2,5):):1:7::]",
+        "[para(2,8):]",
+        "[text(2,8):Test1:]",
+        "[end-para:::True]",
+        "[li(3,5):7::2]",
+        "[para(3,8):]",
+        "[text(3,8):Test2:]",
+        "[end-para:::True]",
+        "[BLANK(4,4):]",
+        "[end-olist:::True]",
+        "[end-block-quote:::True]",
+        "[fcode-block(5,3):`:3:text:::::]",
+        "[text(6,1):block:]",
+        "[end-fcode-block:::3:False]",
+        "[BLANK(8,1):]",
+        "[end-ulist:::True]",
+    ]
+    expected_gfm = """<ul>
+<li>Test List
+<blockquote>
+<ol>
+<li>Test1</li>
+<li>Test2</li>
+</ol>
+</blockquote>
+<pre><code class="language-text">block
+</code></pre>
+</li>
+</ul>"""
+
+    # Act & Assert
+    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+
+
+@pytest.mark.gfm
+@pytest.mark.skip
+def test_nested_three_unordered_block_ordered_with_blank_fenced_c():
+    """
+    TBD - from https://github.com/jackdewinter/pymarkdown/issues/731
+    """
+
+    # Arrange
+    source_markdown = """- Test List
+  > 1) Test1
+  > 2) Test2
+  >
+  ```text
+  block
+  text
+  ```
+"""
+
+    expected_tokens = [
+        "[ulist(1,1):-::2::  \n\n  \n  \n  \n]",
+        "[para(1,3):]",
+        "[text(1,3):Test List:]",
+        "[end-para:::True]",
+        "[block-quote(2,3):  :  > \n  > \n  >]",
+        "[olist(2,5):):1:7::]",
+        "[para(2,8):]",
+        "[text(2,8):Test1:]",
+        "[end-para:::True]",
+        "[li(3,5):7::2]",
+        "[para(3,8):]",
+        "[text(3,8):Test2:]",
+        "[end-para:::True]",
+        "[BLANK(4,4):]",
+        "[end-olist:::True]",
+        "[end-block-quote:::True]",
+        "[fcode-block(5,3):`:3:text:::::]",
+        "[text(6,1):block\ntext:]",
+        "[end-fcode-block:::3:False]",
+        "[BLANK(9,1):]",
+        "[end-ulist:::True]",
+    ]
+    expected_gfm = """<ul>
+<li>Test List
+<blockquote>
+<ol>
+<li>Test1</li>
+<li>Test2</li>
+</ol>
+</blockquote>
+<pre><code class="language-text">block
+text
+</code></pre>
+</li>
+</ul>"""
+
+    # Act & Assert
+    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=True)
 
 
 @pytest.mark.gfm
@@ -1846,11 +1960,10 @@ def test_nested_three_unordered_block_ordered_with_blank_html_x():
 <!-- this is html-->"""
 
     # Act & Assert
-    act_and_assert(source_markdown, expected_gfm, expected_tokens, show_debug=False)
+    act_and_assert(source_markdown, expected_gfm, expected_tokens)
 
 
 @pytest.mark.gfm
-@pytest.mark.skip
 def test_nested_three_unordered_block_ordered_with_blank_html_a():
     """
     TBD - from https://github.com/jackdewinter/pymarkdown/issues/731
@@ -1872,7 +1985,7 @@ def test_nested_three_unordered_block_ordered_with_blank_html_a():
         "[text(1,3):Headline 1: ]",
         "[end-atx::]",
         "[BLANK(2,1):]",
-        "[ulist(3,1):-::2::\n\n]",
+        "[ulist(3,1):-::2::\n  \n\n]",
         "[para(3,3):]",
         "[text(3,3):Test List:]",
         "[end-para:::True]",
@@ -1889,11 +2002,11 @@ def test_nested_three_unordered_block_ordered_with_blank_html_a():
         "[BLANK(7,1):]",
         "[end-olist:::True]",
         "[end-block-quote:::True]",
-        "[end-ulist:::True]",
-        "[html-block(8,1)]",
-        "[text(8,1):<!-- this is html-->:]",
+        "[html-block(8,3)]",
+        "[text(8,3):<!-- this is html-->:]",
         "[end-html-block:::False]",
         "[BLANK(9,1):]",
+        "[end-ulist:::True]",
     ]
     expected_gfm = """<h1>Headline 1</h1>
 <ul>

